dbus/subscription.go

```diff
@@ -94,58 +95,77 @@ func (c *Conn) dispatch() {
+	return c.SubscribeUnitsCustomContext(ctx, interval, 0, func(u1, u2 *UnitStatus) bool { return *u1 != *u2 }, nil)
+}
+
+// Deprecated: use SubscribeUnitsCustomContext instead.
+func (c *Conn) SubscribeUnitsCustom(interval time.Duration, buffer int, isChanged func(*UnitStatus, *UnitStatus) bool, filterUnit func(string) bool) (<-chan map[string]*UnitStatus, <-chan error) {
+	return c.SubscribeUnitsCustomContext(context.Background(), interval, buffer, isChanged, filterUnit)
 }
 
-// SubscribeUnitsCustom is like SubscribeUnits but lets you specify the buffer
+// SubscribeUnitsCustomContext is like SubscribeUnits but lets you specify the buffer
 // size of the channels, the comparison function for detecting changes and a filter
 // function for cutting down on the noise that your channel receives.
-func (c *Conn) SubscribeUnitsCustom(interval time.Duration, buffer int, isChanged func(*UnitStatus, *UnitStatus) bool, filterUnit func(string) bool) (<-chan map[string]*UnitStatus, <-chan error) {
+func (c *Conn) SubscribeUnitsCustomContext(ctx context.Context, interval time.Duration, buffer int, isChanged func(*UnitStatus, *UnitStatus) bool, filterUnit func(string) bool) (<-chan map[string]*UnitStatus, <-chan error) {
 	old := make(map[string]*UnitStatus)
 	statusChan := make(chan map[string]*UnitStatus, buffer)
 	errChan := make(chan error, buffer)
 
-	go func() {
-		for {
-			timerChan := time.After(interval)
-
-			units, err := c.ListUnits()
-			if err == nil {
-				cur := make(map[string]*UnitStatus)
-				for i := range units {
-					if filterUnit != nil && filterUnit(units[i].Name) {
-						continue
-					}
-					cur[units[i].Name] = &units[i]
+	sendSubscriptionUnitsUpdates := func() {
+		units, err := c.ListUnitsContext(ctx)
+		if err == nil {
+			cur := make(map[string]*UnitStatus)
+			for i := range units {
+				if filterUnit != nil && filterUnit(units[i].Name) {
+					continue
 				}
+				cur[units[i].Name] = &units[i]
+			}
 
-				// add all new or changed units
-				changed := make(map[string]*UnitStatus)
-				for n, u := range cur {
-					if oldU, ok := old[n]; !ok || isChanged(oldU, u) {
-						changed[n] = u
-					}
-					delete(old, n)
+			// add all new or changed units
+			changed := make(map[string]*UnitStatus)
+			for n, u := range cur {
+				if oldU, ok := old[n]; !ok || isChanged(oldU, u) {
+					changed[n] = u
 				}
+				delete(old, n)
+			}
 
-				// add all deleted units
-				for oldN := range old {
-					changed[oldN] = nil
-				}
+			// add all deleted units
+			for oldN := range old {
+				changed[oldN] = nil
+			}
 
-				old = cur
+			old = cur
 
-				if len(changed) != 0 {
-					statusChan <- changed
-				}
-			} else {
-				errChan <- err
+			if len(changed) != 0 {
+				statusChan <- changed
 			}
+		} else {
+			errChan <- err
+		}
+	}
 
-			<-timerChan
+	go func() {
+		// Make the initial check for the unit statuses
+		sendSubscriptionUnitsUpdates()
```

Comment:
I don't follow why this was changed to an extra function. This makes a larger diff than necessary.

All you would need to chnage in the original code is use ListUnitsContext() and then where we wait for timerChan put the select which also waits for the Context and returns